src/test/xmeta/testxmetaemit.cpp

```diff
@@ -0,0 +1,120 @@
+#include "pch.h"
+#include <iostream>
+
+#include "antlr4-runtime.h"
+#include "XlangParser.h"
+#include "XlangLexer.h"
+#include "xlang_model_walker.h"
+#include "xmeta_emit.h"
+#include "xmeta_models.h"
+#include "xlang_test_listener.h"
+#include "meta_reader.h"
+#include "meta_writer.h"
+#include "xmeta_idl_reader.h"
+
+using namespace antlr4;
+
+#pragma execution_character_set("utf-8")
+
+using namespace xlang::xmeta;
+using namespace xlang::meta::reader;
+using namespace std;
+
+// In-depth checking of type properties of enums
+const TypeAttributes enum_type_attributes()
+{
+    TypeAttributes result{};
+    result.Visibility(TypeVisibility::Public);
+    result.Sealed(true);
+    result.WindowsRuntime(true);
+    return result;
+}
+
+TEST_CASE("Assemblies metadata") 
+{
+    std::string assembly_name = "testidl";
+    std::string common_assembly_ref = "mscorlib";
+    std::shared_ptr<xlang::xmeta::xmeta_emit> emitter = std::make_shared<xlang::xmeta::xmeta_emit>(assembly_name);
+    emitter->initialize();
+
+    xlang::meta::writer::pe_writer writer;
+    writer.add_metadata(emitter->save_to_memory());
+
+    xlang::meta::reader::database db{ writer.save_to_memory() };
+
+    REQUIRE(db.Assembly.size() == 1);
+    REQUIRE(db.Assembly[0].Name() == "testidl");
+    REQUIRE(db.AssemblyRef.size() == 1);
+    REQUIRE(db.AssemblyRef[0].Name() == "mscorlib");
+    emitter->uninitialize();
+}
+
+TEST_CASE("Enum metadata")
+{
+    std::istringstream test_idl{ R"(
+        namespace Windows.Test
+        {
+            enum Color
+            {
+                Red,
+                Green,
+                Blue
+            }
+        }
+    )" };
+    std::string assembly_name = "testidl";
+    xmeta_idl_reader reader{ "" };
+    REQUIRE(reader.read(test_idl) == 0);
+    xlang::xmeta::xlang_model_walker walker(reader.get_namespaces());
+
+    std::shared_ptr<xlang::xmeta::xmeta_emit> emitter = std::make_shared<xlang::xmeta::xmeta_emit>(assembly_name);
```

Comment:
Why is the emitter in a shared_ptr?

Reply:
In this case, this test is using xlang_model_walker which takes in a shared pointer for the listener. 

void xlang_model_walker::register_listener(std::shared_ptr<xlang_model_listener> const& listener)

Reply:
And the point of my question is: why is the model walker using a `shared_ptr`? Is this really a case where we expect to have multiple walkers sharing ownership of a listener?